server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java

```diff
@@ -83,4 +89,71 @@ protected void addAccountConsoleClient(RealmModel realm) {
             client.addProtocolMapper(audienceMapper);
         }
     }
+
+
+    protected void migrateAuthenticationFlowsWithAlternativeRequirements(RealmModel realm) {
+        for (AuthenticationFlowModel flow : realm.getAuthenticationFlows()) {
+
+            // Suffix used just to avoid name conflicts
+            int suffix = 0;
+            boolean requiredFound = false;
+            LinkedList<AuthenticationExecutionModel> alternativesAfterRequired = new LinkedList<>();
+            for (AuthenticationExecutionModel execution : realm.getAuthenticationExecutions(flow.getId())) {
+                if (AuthenticationExecutionModel.Requirement.REQUIRED.equals(execution.getRequirement())) {
+                    requiredFound = true;
+                }
+
+                // If we have some REQUIRED then ALTERNATIVE and then REQUIRED/CONDITIONAL, we migrate the alternatives to the new subflow.
+                if (AuthenticationExecutionModel.Requirement.REQUIRED.equals(execution.getRequirement()) ||
+                        AuthenticationExecutionModel.Requirement.CONDITIONAL.equals(execution.getRequirement())) {
+                    if (!alternativesAfterRequired.isEmpty()) {
+                        migrateAlternatives(realm, flow, alternativesAfterRequired, suffix);
+                        suffix += 1;
+                        alternativesAfterRequired.clear();
+                    }
+                }
+
+                // ALTERNATIVE executions found after REQUIRED. We may migrate them
+                if (AuthenticationExecutionModel.Requirement.ALTERNATIVE.equals(execution.getRequirement()) && requiredFound) {
```

Comment:
For a flow like...
- ExecutionA, alternative
- ExecutionB, alternative
- ExecutionC, required
...there would be no migration either. But effectively, this flow is now ignoring step 1 and 2 as no migration is applied (see mail thread on dev mailing list -- our custom SPI use case).

Reply:
You're right. Will improve that.